src/software/ai/navigator/navigator.h

```diff
@@ -1,42 +1,154 @@
-        const World &world,
-        const std::vector<std::unique_ptr<Intent>> &assignedIntents) = 0;
+    void visit(const MoveIntent &move_intent) override;
 
-    virtual ~Navigator() = default;
+    /**
+     * Visits a MoveSpinIntent to perform an operation.
+     *
+     * @param move_spin_intent The MoveSpinIntent to visit
+     */
+    void visit(const MoveSpinIntent &move_spin_intent) override;
+
+    /**
+     * Visits a PivotIntent to perform an operation.
+     *
+     * @param pivot_intent The PivotIntent to visit
+     */
+    void visit(const PivotIntent &pivot_intent) override;
+
+    /**
+     * Visits a StopIntent to perform an operation.
+     *
+     * @param stop_intent The StopIntent to visit
+     */
+    void visit(const StopIntent &stop_intent) override;
+
+   private:
+    /**
+     * Create an obstacle for the given avoid area, with a buffer such that the edge
+     * of the robot does not protrude into the area
+     *
+     * @param avoid_area The area to convert into an obstacle
+     *
+     * @return A obstacle representing the given area
+     */
+    std::optional<Obstacle> obstacleFromAvoidArea(AvoidArea avoid_area);
+
+    // This navigators knowledge / state of the world
+    World world;
+
+    // The current Primitive the navigator has created from an Intent.
+    // This variable is set by each `visit` function
+    std::unique_ptr<Primitive> current_primitive;
+
+    // The current Robot the navigator has navigated for from an Intent.
+    // This variable is set by each `visit` function
+    std::optional<Robot> current_robot;
+
+    // The current destination the navigator has navigated to from an Intent.
+    // This variable is set by each `visit` function
+    Point current_destination;
+
+    std::vector<Obstacle> velocity_obstacles;
+
+    std::vector<std::vector<Point>> planned_paths;
 
     /**
-     * Returns the paths planned for each robot by the navigator
+     * Creates a list of obstacles to avoid based on avoid areas,
+     * enemy team, and friendly team, while excluding the robot attached to robot_id
+     *
+     * @param avoid_areas specifies areas to create obstacles for
+     * @param robot_id current robot that is not an obstacle
+     *
+     * @returns list of obstacles
+     */
+    std::vector<Obstacle> createCurrentObstacles(
+        const std::vector<AvoidArea> &avoid_areas, int robot_id);
+
+    /**
+     * Calculates a factor for how close p is to an enemy obstacle.
+     * 0 = touching or inside
+     * 1 = greater than/equal to 2m away
+     * scaled linearly between these values
+     *
+     * @param p point to evaluate
      *
-     * @return the paths planned for each robot by the navigator
+     * @return A factor from 0 to 1 for how close p is to an enemy obstacle
      */
-    virtual std::vector<std::vector<Point>> getPlannedPaths() = 0;
+    double getCloseToEnemyObstacleFactor(Point &p);
```

Comment:
General thought, we don't need to do it here but this function seems a bit arbitrary.

Reply:
That will be part of Navigator clean up